[jit] Make NoneType <: Optional[T]

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25447 Use static type information to restore type tags
* #25439 Add ivalue::type(), part 1
* #25418 Fix BUILD TEST
* **#25361 [jit] Make NoneType <: Optional[T]**

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

Differential Revision: [D17103072](https://our.internmc.facebook.com/intern/diff/D17103072)

zdevito committed:
[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

zdevito committed:
Update on "[jit] Make NoneType <: Optional[T]"

[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

gh-metadata: pytorch pytorch 25361 gh/zdevito/98/head

zdevito committed:
Update on "[jit] Make NoneType <: Optional[T]"

[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

gh-metadata: pytorch pytorch 25361 gh/zdevito/98/head

zdevito committed:
Update on "[jit] Make NoneType <: Optional[T]"

[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

gh-metadata: pytorch pytorch 25361 gh/zdevito/98/head

zdevito committed:
Update on "[jit] Make NoneType <: Optional[T]"

[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

gh-metadata: pytorch pytorch 25361 gh/zdevito/98/head

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1266,17 +1268,36 @@ CAFFE2_API bool isSubvalueOf(const IValue& input_ivalue, TypePtr type);
+    return MatchTypeReturn();
+  }
+  bool success() const {
+    return reason_ == "";
```
Make `reason_` an optional to avoid false success if you report a failure with an empty string for a reason?

suo commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -465,39 +465,49 @@ static value_list getReverseCaptures(Gradient& grad_desc) {
-        node->replaceInputWith(input, lifted_constant->output());
-      }
-    }
+    if (inBlock(input->node(), if_not_in_this_block))
```
Can you add a comment explaining what's going on here? It's hard for me to understand what this code does

suo commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -919,36 +919,6 @@ struct PythonPrintPass {
     stmt << ss.str();
   }
 
-  void printNone(TaggedStringStream& stmt, const Node* node) {
```
wooo!

suo reviewed (approved):
lgtm

zdevito committed:
Update on "[jit] Make NoneType <: Optional[T]"

[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

gh-metadata: pytorch pytorch 25361 gh/zdevito/98/head

zdevito committed:
Update on "[jit] Make NoneType <: Optional[T]"

[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

gh-metadata: pytorch pytorch 25361 gh/zdevito/98/head

temp

zdevito committed:
Update on "[jit] Make NoneType <: Optional[T]"

[jit] Make NoneType <: Optional[T]

Previously we had a different None object for each type T so that
unwrap optional could still recover the type T from it. After a few
months of having this conversion behavior, it has become clear that
only the unwrap optional operators cause this problem. Furthermore, it
is beneficial to have NoneType <: Optional[T] because this is how IValues
work (in particular the None IValue is not tagged). This patch makes the
necessary changes to do this. In particular it special cases unwrap optional
in export so that it annotates the None to make sure we can recover the type.

This also changes how matching and evaluating type values work so that we
can consider None matchable to type Optional[T], eventhough we cannot
derive T from that match.

gh-metadata: pytorch pytorch 25361 gh/zdevito/98/head

temp
